### PR TITLE
Use email_validator to validate email addresses

### LINKED ISSUE #109: EMAIL_REGEX does not match email with internationalized domain name.
For example, `a@vidéotron.ca` is considered an invalid email address, while `a@videotron.ca` is considered valid. Note that the `validate_email` module accepts both as valid, which is the correct behaviour.

### PR DESCRIPTION
**Pull request by petroswork**
fix #109

### CODE REVIEW

**Comment by petroswork** [comment]
I don't know if github allows to squash the extra commits.

File: pydantic/utils.py [hunk 1]
```diff
@@ -1,16 +1,10 @@
 import re
+import email_validator
```
Comment (samuelcolvin) [inline pydantic/utils.py]: I think we should catch the `ImportError` if `email_validator` isn't installed and either:
* raise an error if `validate_email` is called
* or, fallback to `EMAIL_REGEX`

File: requirements.txt [hunk 2]
```diff
@@ -4,3 +4,4 @@
 
 msgpack-python==0.4.8
 ujson==1.35
+email_validator
```
Comment (samuelcolvin) [inline requirements.txt]: requirements should always be pinned with a version.

File: pydantic/utils.py [hunk 3]
```diff
@@ -1,16 +1,10 @@
...
 
-# max length for domain name labels is 63 characters per RFC 1034
-EMAIL_REGEX = re.compile(
-    r'[^\s@\u0000-\u0020"\'`,]+@'
-    r'(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z0-9]{2,63})',
-    re.I
-)
-
 
 def validate_email(value) -> Tuple[str, str]:
```
Comment (samuelcolvin) [inline pydantic/utils.py]: The docstring below needs to be be updated.

File: tests/test_utils.py [hunk 4]
```diff
@@ -50,9 +49,7 @@ def test_address_valid(value, name, email):
     '\u001f@example.com',
     '"@example.com',
     '\"@example.com',
-    '`@example.com',
     ',@example.com',
-    'foobar <foobar`@example.com>',
```
Comment (samuelcolvin) [inline tests/test_utils.py]: surely this case should still fail? If not can you add a fail with with `Name <email@address>` format.

File: tests/test_utils.py [hunk 5]
```diff
@@ -50,9 +49,7 @@ def test_address_valid(value, name, email):
     '\u001f@example.com',
     '"@example.com',
     '\"@example.com',
-    '`@example.com',
```
Comment (samuelcolvin) [inline tests/test_utils.py]: is this really a valid address?

File: requirements.txt (hunk 2 above)
Comment (samuelcolvin) [reply requirements.txt]: You also need to add `email_validator` to setup.py and add documentation both:
* on installing with `email_validator` (I think `email_validator` should be an optional dependency)
* and, a note saying email validator is used by `validate_email`

File: pydantic/utils.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/utils.py]: You also need to correct import order.

Comment (petroswork) [reply pydantic/utils.py]: If we create a branch with `EMAIL_REGEX`, we'll have to create a branch in the unit tests, too. I don't think it is a good idea to consider an email address valid or not, depending on the modules installed.

File: tests/test_utils.py (hunk 5 above)
Comment (petroswork) [reply tests/test_utils.py]: Sure, according to https://en.wikipedia.org/wiki/Email_address this is allowed, but in practice many MTAs restrict usage to much fewer special characters.

File: pydantic/utils.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/utils.py]: Ok, I still don't want to make `email_validator` an install requirement of pydantic, but I agree that different behaviour depending on whether the package is installed would confuse people. Therefore do something like

```py
 def validate_email(value) -> Tuple[str, str]:
    try:
        import email_validator
    except ImportError as e:
        raise ImportError('email_validator need to be installed to in order to use pydantic for email validation, please run `pip install email_validator`') from e
    ...
```

File: tests/test_utils.py (hunk 5 above)
Comment (samuelcolvin) [reply tests/test_utils.py]: ok

**Comment by samuelcolvin** [comment]
Thanks a lot for this, I've replaced it with #142 as I wanted to start using this and there was quite a lot to change here.

Still this effort was much appreciated.

**Pull request closed without merging**